Add support for chat-style prompts with structured output with prompt object

harshilprajapati96 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/harshilprajapati96/mlflow/pull/16341?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16341/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16341/merge#subdirectory=skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/16341/merge
```

</p>
</details>

Signed-off-by: Harshil Prajapati <47331144+harshilprajapati96@users.noreply.github.com>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #16259 

### What changes are proposed in this pull request?

<!-- Please fill in changes proposed in this PR. -->

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

harshilprajapati96 commented:
@B-Step62 @rohitarun-db any comments? I would really like to get this feature available, let me know if there is any feedback. Happy to work to get it done!

TomeHirata commented on docs/docs/genai/prompt-version-mgmt/prompt-registry/create-and-edit-prompts.mdx:
```diff
@@ -26,6 +28,9 @@ You can initiate a new prompt in the MLflow Prompt Registry in two primary ways:
 3.  You will be prompted to provide:
     - **Name**: A unique name for your new prompt (e.g., `customer-service-greeting`).
     - **Template**: The initial text of your prompt. You can include variables using the `{{variable_name}}` syntax (e.g., `Hello {{customer_name}}, how can I assist you today?`).
+    - **Prompt Type**: Choose between "text" (default) or "chat" for structured conversation prompts.
```
Do we need this argument? We can check the type of `Template`

harshilprajapati96 replied:
will remove

TomeHirata commented on docs/docs/genai/prompt-version-mgmt/prompt-registry/create-and-edit-prompts.mdx:
```diff
@@ -26,6 +28,9 @@ You can initiate a new prompt in the MLflow Prompt Registry in two primary ways:
     - **Template**: The initial text of your prompt. You can include variables using the `{{variable_name}}` syntax (e.g., `Hello {{customer_name}}, how can I assist you today?`).
+    - **Prompt Type**: Choose between "text" (default) or "chat" for structured conversation prompts.
+    - **Response Format**: (Optional) Define a structured output format using JSON schema or Pydantic models.
+    - **Model Configuration**: (Optional) Specify model parameters like temperature, max tokens, etc.
```
IMO, this is out of scope of `prompt`, and should be configured with models. What do you think? @B-Step62

harshilprajapati96 replied:
Other open-source frameworks do have config which can be optional but this makes it usable without loading the whole model. 

ref: https://langfuse.com/docs/prompts/get-started#langfuse-prompt-object

TomeHirata replied:
Langfuse includes config in the prompt data model because they don't have an entity that represents Model. Conversely, in the mlflow data model, parameters such as LLM name or temperature fits more for Logged Model which allows users to reuse the same prompt with different parameters.

harshilprajapati96 replied:
sounds good!

TomeHirata commented on mlflow/tracking/client.py:
```diff
@@ -442,7 +460,10 @@ def create_run(
         self,
         name: str,
-        template: str,
+        template: Union[str, list[dict[str, str]]],
```
Let's use [ContentType](https://github.com/mlflow/mlflow/blob/85a622ab5db81e6b38848f93fc1de12cb872401d/mlflow/types/chat.py#L72)

TomeHirata commented on mlflow/tracking/client.py:
```diff
@@ -442,7 +460,10 @@ def create_run(
         name: str,
-        template: str,
+        template: Union[str, list[dict[str, str]]],
+        prompt_type: str = "text",
```
IMO, we don't need `prompt_type` and `config`. For this PR, can we focus on multi-turn template and `response_format`?

harshilprajapati96 replied:
sure, I am okay with that I can store config in tags but having config would make sense.

TomeHirata commented on mlflow/tracking/client.py:
```diff
@@ -512,6 +556,18 @@ def register_prompt(
+        # Auto-detect prompt type based on template structure
+        if prompt_type == "text" and isinstance(template, list):
+            # Check if it's a valid chat template (list of dicts with role and content)
+            if len(template) > 0 and all(
```
We can use `ContentPartsList.model_validate`

TomeHirata commented on mlflow/tracking/client.py:
```diff
@@ -563,7 +624,36 @@ def register_prompt(
+            if isinstance(response_format, type):
+                # Convert Pydantic class to JSON schema
+                try:
+                    from pydantic import BaseModel
```
Since pydantic is a core dependency, let's import in the top level

TomeHirata commented on mlflow/tracking/client.py:
```diff
@@ -2884,9 +2980,8 @@ def log_image(
-            uncompressed_filename = (
-                f"images/{sanitized_key}%step%{step}%timestamp%{timestamp}%{filename_uuid}"
-            )
+            uncompressed_filename_prefix = f"images/{sanitized_key}%step%{step}%timestamp%"
```
Why was this part changed?

TomeHirata commented on mlflow/tracking/client.py:
```diff
@@ -496,9 +538,11 @@ def register_prompt(
+            template: The prompt template. For text prompts, a string with variables in {{}}.
+                     For chat prompts, a list of message dictionaries with 'role' and 'content'.
+            prompt_type: The type of prompt ("text" or "chat"). Defaults to "text".
+            response_format: Optional Pydantic class or dict defining the response structure.
```
What is the expected format of `dict`? Do we expect a full json schema?

harshilprajapati96 replied:
yeah json schema like https://github.com/openai/openai-python/blob/1dbb72b290cb360a9ccbcec17425aaba4ad114b0/src/openai/lib/_parsing/_completions.py#L232

TomeHirata commented on mlflow/store/model_registry/abstract_store.py:
```diff
@@ -701,8 +712,41 @@ def create_prompt_version(
+            version_tags.append(ModelVersionTag(key=PROMPT_TYPE_TAG_KEY, value=prompt_type))
+
+        if response_format:
+            if isinstance(response_format, type):
```
ditto

harshilprajapati96 commented on mlflow/store/model_registry/abstract_store.py:
```diff
@@ -701,8 +712,41 @@ def create_prompt_version(
+                    format_dict = {"type": "object", "properties": {}}
+            else:
+                format_dict = response_format
+            version_tags.append(
```
Not sure what you mean? I just need pydantic/json schema stored as part of prompt so that we can use as part of request and we can see in catalog what is the expected format

TomeHirata replied:
I was just wondering if automatically converting pydantic to json schema is an appropriate behavior here. But I agree this is ok.

harshilprajapati96 committed:
Enhance prompt registration and formatting capabilities

- Introduced support for chat prompts with structured message arrays and added response format specifications.
- Updated documentation to reflect new features, including examples for both text and chat prompts.
- Improved the PromptVersion class to handle different template types and response formats effectively.

(cherry picked from commit 896d504db7b6e9e46ad4a047dd7547487f71a6e4)
Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 commented:
@TomeHirata Updated the code 
- removed config
- auto-detect template
- only support chat style & support response format

harshilprajapati96 committed:
Refactor prompt documentation and enhance structured output support

- Updated the prompts documentation to clarify the use of tags and response formats.
- Introduced a new section on structured output, detailing how to define and validate response formats using Pydantic models and JSON schemas.
- Improved the PromptVersion class to cache deserialized templates and response formats for better performance.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 commented on mlflow/entities/model_registry/prompt_version.py:
```diff
@@ -11,14 +14,24 @@
+    }
 
 
 class PromptVersion(_ModelRegistryEntity):
```
I can have a separate draft PR for the refactor and we can discuss if that's the right direction? Can we please get this deployed?

harshilprajapati96 committed:
revert docs/docs/genai/prompt-version-mgmt/prompt-registry/create-and-edit-prompts.mdx changes

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 commented:
@B-Step62 addressed all comments! Thanks!

TomeHirata commented:
@harshilprajapati96 Thank you for addressing the comments. The code overall looks good, can you fix the doc failures and test failures?

harshilprajapati96 committed:
Refactor PromptVersion class and clean up prompt registration

- Removed caching for deserialized templates and response formats in the PromptVersion class to simplify the code.
- Streamlined the JSON parsing process for prompt text and response format retrieval.
- Minor formatting adjustments in the register_prompt function for improved readability.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 committed:
Refactor prompt documentation and streamline prompt registration

- Simplified the prompt registration examples by removing chat prompt references and focusing on text prompts.
- Updated the template structure to clarify variable usage and response format specifications.
- Enhanced documentation consistency by aligning terminology and structure across sections.
- Removed redundant checks and examples for chat prompts to streamline the content.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 committed:
Update prompt documentation to reflect class name change from `PromptVersion` to `Prompt`.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 committed:
✅ Fixed: WARNING: Unexpected indentation.
✅ Fixed: WARNING: Block quote ends without a blank line; unexpected unindent.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 commented:
> @harshilprajapati96 Thank you for addressing the comments. The code overall looks good, can you fix the doc failures and test failures?

Done

TomeHirata commented:
It seems docs ci is still failing, can you run `yarn prettier:fix` under docs/?

harshilprajapati96 committed:
fix formatting

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 commented:
>Also the skinny test is failing due to a numpy dependency error.
should I add it to skinny? won't it make the package huge? Any suggestions?

B-Step62 commented:
> should I add it to skinny? won't it make the package huge? Any suggestions?

@harshilprajapati96 The root cause is that `mlflow.types.chat` is imported at the global level in `mlflow/entities/model_registry/prompt_version.py`. We should move the import to inside the function to make it lazy load. 

For `ContentType`, you can use `TYPE_CHECKING` pattern like [this example](https://github.com/mlflow/mlflow/blob/dca6eaf1423a37590b4c535b1cefa7f80613e8d8/mlflow/genai/datasets/evaluation_dataset.py#L11) to avoid importing it at runtime.

harshilprajapati96 committed:
Refactor type hinting in PromptVersion class

- Updated the import of ContentType to be conditional based on TYPE_CHECKING for improved type hinting.
- Adjusted the template parameter type hint to use a string literal for ContentType, enhancing clarity.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

TomeHirata commented on mlflow/entities/model_registry/prompt_version.py:
```diff
@@ -191,21 +296,38 @@ def _properties(cls) -> list[str]:
-    def format(self, allow_partial: bool = False, **kwargs) -> Union[PromptVersion, str]:
+    def format(
+        self, allow_partial: bool = False, **kwargs
+    ) -> Union[PromptVersion, str, list[dict[str, ContentType]]]:
```
Shouldn't this be `Union[PromptVersion, ContentType]`?

TomeHirata replied:
Sorry, you're right. Let's keep it as it is.

harshilprajapati96 committed:
Refactor type hinting for improved clarity and consistency

- Updated type hints in the PromptVersion class and related functions to use Optional and Union more effectively.
- Changed the import of ContentType to be conditional based on TYPE_CHECKING across multiple files for better type hinting.
- Enhanced the template parameter type hint to use a string literal for ContentType, improving code readability.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

TomeHirata commented:
Hi, @harshilprajapati96. Thanks for the updates. We hope to merge this PR soon. Can you fix the doc errors? Alternatively, would you mind my adding some commits to this PR?

harshilprajapati96 committed:
Enhance prompt registration documentation with structured templates

- Updated the prompt registration examples to include structured templates for both initial and new versions.
- Introduced a Pydantic model for response formats to improve clarity on expected outputs.
- Enhanced the documentation to provide clearer instructions on using tags and commit messages during prompt registration.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

harshilprajapati96 commented:
> Hi, @harshilprajapati96. Thanks for the updates. We hope to merge this PR soon. Can you fix the doc errors? Alternatively, would you mind my adding some commits to this PR?

done, sorry about that!

TomeHirata commented:
Can you make this change?

```diff
diff --git a/mlflow/entities/model_registry/prompt_version.py b/mlflow/entities/model_registry/prompt_version.py
index 3f616f4d0..4d5ef0fb3 100644
--- a/mlflow/entities/model_registry/prompt_version.py
+++ b/mlflow/entities/model_registry/prompt_version.py
@@ -70,13 +70,13 @@ class PromptVersion(_ModelRegistryEntity):
         name: str,
         version: int,
         template: Union[str, list[dict[str, "ContentType"]]],
-        response_format: Optional[Union[BaseModel, dict[str, Any]]] = None,
         commit_message: Optional[str] = None,
         creation_timestamp: Optional[int] = None,
         tags: Optional[dict[str, str]] = None,
         aliases: Optional[list[str]] = None,
         last_updated_timestamp: Optional[int] = None,
         user_id: Optional[str] = None,
+        response_format: Optional[Union[BaseModel, dict[str, Any]]] = None,
     ):
         from mlflow.types.chat import ChatMessage
 
diff --git a/mlflow/types/__init__.py b/mlflow/types/__init__.py
index 9f3da7db8..64d42f8ff 100644
--- a/mlflow/types/__init__.py
+++ b/mlflow/types/__init__.py
@@ -3,9 +3,9 @@ The :py:mod:`mlflow.types` module defines data types and utilities to be used by
 components to describe interface independent of other frameworks or languages.
 """
 
-from mlflow.version import IS_TRACING_SDK_ONLY
+from mlflow.version import IS_FULL_MLFLOW
 
-if not IS_TRACING_SDK_ONLY:
+if IS_FULL_MLFLOW:
     import mlflow.types.llm  # noqa: F401
 
     # Our typing system depends on numpy, which is not included in mlflow-tracing package
diff --git a/mlflow/version.py b/mlflow/version.py
index e17670c06..7421bb1bd 100644
--- a/mlflow/version.py
+++ b/mlflow/version.py
@@ -22,3 +22,4 @@ def _is_package_installed(package_name: str) -> bool:
 # This is used to determine whether to import modules that require
 # dependencies that are not included in the tracing SDK.
 IS_TRACING_SDK_ONLY = not any(_is_package_installed(pkg) for pkg in ["mlflow", "mlflow-skinny"])
+IS_FULL_MLFLOW = any(_is_package_installed(pkg) for pkg in ["mlflow"])
```

B-Step62 commented:
@harshilprajapati96 We aim to release the feature in 3.2 RC next Monday. Would it be possible for you to address CI failure or give us permission to directly push fixes onto your branch?

harshilprajapati96 committed:
Update version handling and enhance PromptVersion class

- Added a new constant IS_FULL_MLFLOW to determine if the full MLflow package is installed.
- Adjusted the import logic in mlflow.types to conditionally import mlflow.types.llm based on the new constant.
- Moved the response_format parameter in the PromptVersion class to improve clarity and consistency in type hinting.

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

TomeHirata commented:
Hi, @harshilprajapati96. There was a test only issue due to the sys path modification of pytest. Can you apply following change?

```diff
diff --git a/dev/run-python-skinny-tests.sh b/dev/run-python-skinny-tests.sh
index cda37eb36..7520741c6 100755
--- a/dev/run-python-skinny-tests.sh
+++ b/dev/run-python-skinny-tests.sh
@@ -36,7 +36,8 @@ pytest \
   tests/utils/test_requirements_utils.py::test_infer_requirements_excludes_mlflow \
   tests/utils/test_search_utils.py \
   tests/store/tracking/test_file_store.py \
-  tests/utils/test_doctor.py
+  tests/utils/test_doctor.py \
+  --import-mode=importlib
 
 python -m pip install pandas
 pytest tests/test_skinny_client_autolog_without_scipy.py
diff --git a/mlflow/version.py b/mlflow/version.py
index 7421bb1bd..59859b412 100644
--- a/mlflow/version.py
+++ b/mlflow/version.py
@@ -22,4 +22,4 @@ def _is_package_installed(package_name: str) -> bool:
 # This is used to determine whether to import modules that require
 # dependencies that are not included in the tracing SDK.
 IS_TRACING_SDK_ONLY = not any(_is_package_installed(pkg) for pkg in ["mlflow", "mlflow-skinny"])
-IS_FULL_MLFLOW = any(_is_package_installed(pkg) for pkg in ["mlflow"])
+IS_FULL_MLFLOW = _is_package_installed("mlflow")
```

harshilprajapati96 committed:
Update test script and version handling

- Added the `--import-mode=importlib` flag to the test script for improved module import handling.
- Simplified the logic for determining if the full MLflow package is installed by directly calling `_is_package_installed` with "mlflow".

Signed-off-by: harshilprajapati96 <harshil.prajapati@coalitioninc.com>

TomeHirata reviewed (approved):
Overall, LGTM, thank you for the great addition! 
I'll follow up on some nit fixes in a separate PR with the interest of time.
